doc/InternalRep.3
'\"
'\" Copyright (c) 1996-1998 Sun Microsystems, Inc.
'\"
'\" See the file "license.terms" for information on usage and redistribution
'\" of this file, and for a DISCLAIMER OF ALL WARRANTIES.
'\" 
'\" RCS: @(#) $Id$
'\" 
.so man.macros
.TH "InternalRep" 3 1.0 Jacl "Jacl Library Procedures"
.BS
.SH CLASS
tcl.lang.InternalRep -- This class is an abstract base class of all
the internal representation types in Tcl.

.SH METHODS
.sp
void \fBdispose\fR()
.sp
InternalRep \fBduplicate\fR()
.sp
String \fBtoString\fR()

.BE
.SH DESCRIPTION
.PP
The \fBInternalRep\fR class is an abstract base class of all the
internal representation types in Tcl.  All InternalRep subclasses must
implement the following three methods.
.TP
\fBdispose\fR
The \fBdispose\fR method is called when an InternalRep is no longer
used by Tcl. It should free all the resources used by this InternalRep
instance.
.TP
\fBduplicate\fR
The \fBduplicate\fR method is called when an exact copy of the
InternalRep is needed during a copy-on-write operation.
.TP
\fBtoString\fR
The \fBtoString\fR method is called when a string representation is
needed for a TclObject.
.SH "EQUIVALENT C FUNCTIONS"
Tcl_ObjType.freeProc, Tcl_ObjType.dupIntRepProc, Tcl_ObjType.updateStringProc
.SH "SEE ALSO"
TclObject, TclList, TclIndex, TclDouble, TclString, TclInteger, TclBoolean, ReflectObject
.SH KEYWORDS
internal representation, object, object type, abstract class


ChangeLog
2019-05-04  Fabrice Ducos <fabrice.ducos@gmail.com>

        * Support of latest versions of Tcl (at least Tcl 8.5
        and Tcl 8.6) and of modern JDK (10+);
        support of JDK 4- abandoned (use of auto-boxing);
        support of OSX systems added;
        details below

        * configuration (tcljava.m4, configure.in),
        and compilation (Makefile.in) templates updated for
        JDK 10+ (javah not available anymore).
        javah is still used when it is available (JDK 9-).
        * minor update of jtclsh.in (for OSX dynamic libraries)
        * many obsolescent constructs (mostly old-style boxing)
        removed in *.java under src/jacl/tcl/lang,
        src/tclblend/tcl/lang and src/tcljava/tcl/lang,
        in order to silence many warnings.
        (effect: support of JDK 4- abandoned)
        * src/native/javaInterp.c: 2010 implementation
        of Java_tcl_lang_Interp_initName() crashed on the
        example given at docs/website/getstart.html,
        because calling Tcl_GetNameOfExecutable
        before Tcl_FindExecutable gives rise to undefined
        behaviour, at least in latest versions of Tcl
        (crashes observed on several systems, at least with
        Tcl 8.6).
        Java_tcl_lang_Interp_initName() is now simplified and
        always calls Tcl_FindExecutable first.
        * all the occurences of interp->result have been
        replaced with Tcl_GetStringResult(interp).
        Accessing interp->result directly was deprecated in
        latest versions of the Tcl C API, and removed from
        Tcl 8.6 (another solution was to #define USE_INTERP_RESULT,
        but this would have been an easy-going, quick and dirty way)
        ref. https://www.tcl.tk/man/tcl/TclLib/Interp.htm

2019-05-01  Fabrice Ducos <fabrice.ducos@gmail.com>

        * tcljava resurrected
        * migration from a CVS to a GIT repository
        (with reposurgeon)

2010-02-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* tests/tcl/regexp.test: Add test case for
        escaped - character in regexp command.
        [Bug 2012389]

2010-02-21  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceAll):
	Fixup regsub -all logic so 5 more tests pass.
	The replaceAll() logic is broken because it depends
	on matching substrings instead of using an index
	into the larger string and Java's regexp library
	uses this to calculate line starts in multiline mode.

2010-02-18  Mo DeJong  <mdejong@users.sourceforge.net>

        Incorporate fix/tests from Tcl head on 2010-02-11,
        aka was Tcl [Bug 2826551, Patch 2948425].

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
        Fixup increment index on zero length match issue.
	* tests/tcl/regexp.test: Back port of new regexp tests,
        some tests marked with {endIndex} due to lack of impl.

2010-02-11  Mo DeJong  <mdejong@users.sourceforge.net>

	Patch 2946118

	* src/jacl/tcl/lang/ListCmd.java: Minor style reformat.
	* src/jacl/tcl/lang/Util.java (scanElement, convertElement):
	Remove crusty old code that treated backslash zero
	as the end of the line. That logic was removed from
	C Tcl long ago.
	* tests/inprogress/EventAdaptor.test: Fixup tests
	that depended on backslash zero quirks.
	* tests/tcl/list.test: Add test cases for backslash zero
	as argument to list command.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (match, _match,
        replaceFirst, replaceAll): Rewrite -start offset
        logic to correctly deal with odd edge cases.
	* tests/tcl/regexp.test: Add additional test cases
        that cover -start index edge cases.

2009-10-04  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc):
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Rewrite argument processing logic to avoid using
        IndexOutOfBoundsException to deal with an incorrect
        number of arguments.

2009-10-03  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fix regsub -all crash because offset was not getting
        advanced by 1 on a match of length zero.
        Fixup -start argument support with and without -all
        in the case where the offset was larger than the
        length of the string.
	* tests/tcl/regexp.test: Add additional test cases.
        Enable regsub test cases that were getting caught
        in an infinite loop.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        failing tests related to optional subgroups.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Minor paren fix, no function changes.

2009-09-30  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Rework
        logic related to match result groups.
	* tests/tcl/regexp.test: Add missing test cases for
        non-capturing groups and -inline results.

2009-09-26  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup use of MULTILINE flag.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for -linestop and -lineanchor arguments to regexp.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):  Fixup
        support for -linestop and -lineanchor arguments to regsub.
	* tests/tcl/regexp.test: Add missing test cases for
        -linestop and -lineanchor arguments.

2009-09-22  Mo DeJong  <mdejong@users.sourceforge.net>

	* src/jacl/tcl/lang/Regex.java (replaceFirst, replaceAll):
        Fixup support for the empty string as the string
        argument to regsub by removing the MULTILINE flag when
        an empty string is detected.
	* src/jacl/tcl/lang/RegexpCmd.java (cmdProc): Fixup
        support for empty string as string argument.
	* src/jacl/tcl/lang/RegsubCmd.java (cmdProc):
        Remove comments.
	* tests/tcl/regexp.test: Add additional test cases

src/tcljava/tcl/lang/InternalRep.java
/*
 * InternalRep.java
 *
 *	This file contains the abstract class declaration for the
 *	internal representations of TclObjects.
 *
 * Copyright (c) 1997 Sun Microsystems, Inc.
 *
 * See the file "license.terms" for information on usage and
 * redistribution of this file, and for a DISCLAIMER OF ALL
 * WARRANTIES.
 * 
 * RCS: @(#) $Id: InternalRep.java,v 1.4 1999/06/21 03:32:39 mo Exp $
 *
 */

package tcl.lang;

/**
 * This is the interface for implementing internal representation of Tcl
 * objects.  A class that implements InternalRep should define the
 * following:
 *
 * (1) the two abstract methods specified in this base class:
 *		dispose()
 *		duplicate()
 *
 * (2) The method toString()
 *
 * (3) class method(s) newInstance() if appropriate
 *
 * (4) class method set<Type>FromAny() if appropriate
 *
 * (5) class method get() if appropriate
 */

abstract public class InternalRep {

/*
 *----------------------------------------------------------------------
 *
 * dispose --
 *
 *	Free any state associated with the object's internal rep.
 *
 * Results:
 *	None.
 *
 * Side effects:
 *	Leaves the object in an unusable state.
 *
 *----------------------------------------------------------------------
 */

protected void
dispose()
{
    // The default implementation does nothing.
}

/*
 *----------------------------------------------------------------------
 *
 * duplicate --
 *
 *	Make a copy of an object's internal representation.
 *
 * Results:
 *	Returns a newly allocated instance of the appropriate type.
 *
 * Side effects:
 *	None.
 *
 *----------------------------------------------------------------------
 */

abstract protected InternalRep duplicate();

} // end InternalRep

